[13.0][MIG] ebill_paynet

TDu opened the pull request:
This is based on #499 for 12.0 that has not been merged


- [x] account_invoice_export OCA/edi#212
- [x] base_ebill_payment_contract OCA/edi#207

TDu committed:
[12.0][ADD] ebill_paynet module

TDu committed:
[12.0][FIX] ebill_paynet

TDu committed:
[13.0][MIG] ebill_paynet

TDu committed:
Improvements while testing

TDu committed:
Run pre-commit hooks ebill_paynet

TDu committed:
Add interfacing with account_invoice_export

TDu committed:
Remove compute on user and password

This is removed because there is dedicated addon
`server_env_ebill_payneta` to manage that

TDu committed:
Improve tree view for messages

TDu committed:
[13.0][IMP] Split jobs

Before one job was polling for shipment and downloading/acknowledging
them. This could lead to incoherent state between Odoo and the distant
server.
Now one job polls for shipment and creates one job for each new shipment
to be downloaded and parsed.

Also add improvement in the UI.

sebalix commented:
This depends on some other modules, maybe add that in `oca_dependencies.txt` to get tests running on Travis?

- `account_invoice_export` https://github.com/OCA/edi/pull/212
- `base_ebill_payment_contract` https://github.com/OCA/edi/pull/207

sebalix commented on oca_dependencies.txt:
```diff
@@ -0,0 +1,2 @@
+server-env
+queue
```
you will need `edi` too for https://github.com/OCA/edi/pull/212 and https://github.com/OCA/edi/pull/207.

TDu replied:
Added

sebalix commented on ebill_paynet/components/api.py:
```diff
@@ -0,0 +1,55 @@
+    """PayNet DWS web services."""
+
+    def __init__(self, url, test_service):
+        settings = Settings(xml_huge_tree=True)
```
I would use directly `zeep.Settings`, this avoid to read the imports statements to know what we are talking about as it's not an usual dependency in the Odoo ecosystem. But it's a matter of taste I suppose :)

TDu replied:
Yes clearer, but with an `import zeep` are we not importing more than necessary and it would be bad performance wise ? (real question that I don't know the answer)

yvaucher replied:
@TDu good question to ask.

AFAICT `import` does 2 things, it imports the modules and submodules and fill the namespace.

The best practices says that `from module import *` is discouraged. Because it fills the namespace with every part of  `module`.

When using the `from` you target a submodule or code of the main module in any case you will also load the module itself.

If you test the following in your python shell to load :

```
>>> import sys
>>> sys.modules['zeep']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'zeep'
>>> from zeep import Settings
>>> Settings
<class 'zeep.settings.Settings'>
>>> sys.modules['zeep']
<module 'zeep' from '/home/yvaucher/.local/lib/python3.6/site-packages/zeep/__init__.py'>
>>> sys.modules['zeep'].Settings
<class 'zeep.settings.Settings'>
>>> sys.modules['zeep'].Client
<class 'zeep.client.Client'>
>>> Client
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'Client' is not defined
```

As you can see the `zeep` module is fully loaded but the namespace doesn't contains everything.



But what is a submodule?
https://docs.python.org/3/reference/import.html#submodules

A submodule is a sub file or directory that is not mentionned in the `__init__.py`
It's different from code that is part of the main module which must appear in the `__init__.py` file.



To deep in details, you can't import too much things unless you ask for a submodule you don't use.

Consider the following module with 2 submodules and some initialization code:

```
module
├── initialization.py
├── __init__.py
├── sub1.py
└── sub2.py
```

The content of `module/__init__.py` is 
``` python
from . import initialization

print('import whole module')
```

Now let's test it:

```
>>> import module
load initialization
import whole module
>>> module.initialization
<module 'module.initialization' from '/tmp/module/initialization.py'>
>>> module.sub1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'module' has no attribute 'sub1'
```
Only what is mentionned as `from . import ....`  (in `module/__init__.py` file) is loaded with the module.

-----------------
```
>>> from module import initialization
load initialization
import whole module
>>> from module import sub1
load sub1
>>> from module import sub1, sub2
load sub2
```
Same here, only what is mention as `from . import ....` (in `module/__init__.py` file) is loaded with the module.
You also see that it reminds what has been previously loaded.

-----------------
```
>>> from module import sub1
load initialization
import whole module
load sub1
```

When importing `sub1` the whole module is loaded anyway. But not other sub packages.

-----------------
```
>>> from module import *
load initialization
import whole module
>>> initialization
<module 'module.initialization' from '/tmp/module/initialization.py'>
>>> sub1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'sub1' is not defined
```
You might have expected as I did to see an import of sub1 and sub2 but it's not the case, which means that this is discouraged only due to the namespace filling.


Additional note: If you do want to load a subpackage which has a too generic name you could also define an alias on import using 
`from module import subpackage as msub`
This can save you from namespace conflicts.


In conclusion, the proposal of @sebalix is perfectly fine and has not impact other than having one different entry in the name space.


Some references:

https://docs.python.org/3/reference/import.html
https://docs.python.org/3/library/functions.html#__import__
https://stackoverflow.com/questions/12229580/python-importing-a-sub-package-or-sub-module

sebalix commented on ebill_paynet/components/api.py:
```diff
@@ -0,0 +1,55 @@
+
+    def __init__(self, url, test_service):
+        settings = Settings(xml_huge_tree=True)
+        session = Session()
```
Same with `requests.Session` etc

sebalix commented on ebill_paynet/data/transmit.method.xml:
```diff
@@ -0,0 +1,11 @@
+        <field name="customer_ok">True</field>
+        <!-- There is a bug on transmit_method module, needs to be set -->
+        <!-- to True so it is available for customer -->
+        <field name="supplier_ok">True</field>
```
```suggestion
        <field name="supplier_ok" eval="True"/>
```

TDu replied:
Changed, thanks

sebalix commented on ebill_paynet/data/transmit.method.xml:
```diff
@@ -0,0 +1,11 @@
+    <record id="paynet_transmit_method" model="transmit.method">
+        <field name="name">Paynet</field>
+        <field name="code">paynet</field>
+        <field name="customer_ok">True</field>
```
```suggestion
        <field name="customer_ok" eval="True"/>
```

sebalix commented on ebill_paynet/models/paynet_invoice_message.py:
```diff
@@ -0,0 +1,130 @@
+from ..components.api import PayNetDWS
+
+MODULE_PATH = get_module_root(os.path.dirname(__file__))
+INVOICE_TEMPLATE_2013 = "invoice-2013A.xml"
+INVOICE_TEMPLATE_2003 = "invoice-2003A.xml"
```
As they are Jinja templates:
```suggestion
INVOICE_TEMPLATE_2013 = "invoice-2013A.xml.j2"
INVOICE_TEMPLATE_2003 = "invoice-2003A.xml.j2"
```
(files to rename accordingly)

sebalix commented on ebill_paynet/models/paynet_invoice_message.py:
```diff
@@ -0,0 +1,130 @@
+MODULE_PATH = get_module_root(os.path.dirname(__file__))
+INVOICE_TEMPLATE_2013 = "invoice-2013A.xml"
+INVOICE_TEMPLATE_2003 = "invoice-2003A.xml"
+TEMPLATE_DIR = MODULE_PATH + "/messages"
```
Use `os`:
```suggestion
TEMPLATE_DIR = os.path.join(MODULE_PATH, "messages")
```

sebalix commented on ebill_paynet/models/account_invoice.py:
```diff
@@ -0,0 +1,120 @@
+            return
+        paynet_method = self.env.ref("ebill_paynet.paynet_transmit_method")
+        if self.transmit_method_id == paynet_method:
+            # It is not called when selecting the customer ?
```
Add a FIXME explaining it's due to the `transmit_method_id` added twice in `account_invoice_transmit_method`, it has to be fixed there

https://github.com/OCA/account-invoicing/pull/682/files?file-filters%5B%5D=.csv&file-filters%5B%5D=.html&file-filters%5B%5D=.png&file-filters%5B%5D=.py&file-filters%5B%5D=.rst&file-filters%5B%5D=.xml#r439977482

sebalix commented on ebill_paynet/models/account_invoice.py:
```diff
@@ -0,0 +1,120 @@
+            )
+        return contract
+
+    def create_paynet_message(self):
```
It has to be a private method as you are returning a recordset (not compatible with the JSON/XML-RPC layer):
```suggestion
    def _create_paynet_message(self):
```
(or return the ID of the message record but you'll have to rebrowse it after)

sebalix commented on ebill_paynet/models/account_invoice.py:
```diff
@@ -0,0 +1,120 @@
+        activity = self.activity_reschedule(
+            [activity_type], date_deadline=fields.Date.today()
+        )
+        values = {}
+        if not activity:
+            message = self.env.ref("ebill_paynet.dws_reject_invoice").render(
+                values=values
+            )
```
`values` is empty here, is it now useless?

sebalix commented on ebill_paynet/models/paynet_invoice_message.py:
```diff
@@ -0,0 +1,130 @@
+        ondelete="restrict",
+        readonly=True,
+    )
+    invoice_id = fields.Many2one(comodel_name="account.move", ondelete="restrict")
```
`required=True`?

sebalix commented on ebill_paynet/models/paynet_invoice_message.py:
```diff
@@ -0,0 +1,130 @@
+    )
+    invoice_id = fields.Many2one(comodel_name="account.move", ondelete="restrict")
+    attachment_id = fields.Many2one("ir.attachment", "PDF")
+    state = fields.Selection(
+        selection=[
+            ("draft", "Draft"),
+            ("sent", "Sent"),
+            ("done", "Done"),
+            ("reject", "Reject"),
+            ("error", "Error"),
+        ],
+        default="draft",
+    )
```
`required=True`

sebalix commented on ebill_paynet/models/paynet_invoice_message.py:
```diff
@@ -0,0 +1,130 @@
+    )
+    # Set with invoice_id.number but also with returned data from server ?
+    ref = fields.Char("Reference N°", size=35)
+    ebill_account_number = fields.Char("Paynet Id", size=20)
```
`required=True`?

sebalix commented on ebill_paynet/models/paynet_invoice_message.py:
```diff
@@ -0,0 +1,130 @@
+
+    def _generate_payload(self):
+        self.ensure_one()
+        assert self.state == "draft"
```
A `UserError` is better here (or in `send_to_paynet` method and keep the assert here) as nothing prevent the user to run this (no filter on `draft` state).

sebalix commented on ebill_paynet/models/paynet_invoice_message.py:
```diff
@@ -0,0 +1,130 @@
+                shipment_id = message.service_id.take_shipment(message.payload)
+                message.shipment_id = shipment_id
+                message.state = "sent"
+            except Fault as e:
```
```suggestion
            except zeep.exceptions.Fault as e:
```

sebalix commented on ebill_paynet/models/paynet_service.py:
```diff
@@ -0,0 +1,220 @@
+    _name = "paynet.service"
+    _description = "Paynet service configuration"
+
+    name = fields.Char()
```
Having the name required is a good thing I think

TDu replied:
Agree, done

sebalix commented on ebill_paynet/models/paynet_service.py:
```diff
@@ -0,0 +1,220 @@
+        help="Specify the type of XML exchange with the service.",
+    )
+    partner_bank_id = fields.Many2one(
+        comodel_name="res.partner.bank", string="Bank account",
```
`ondelete="restrict"`?

TDu replied:
Yes, done

sebalix commented on ebill_paynet/models/paynet_service.py:
```diff
@@ -0,0 +1,220 @@
+                Authorization=dws.authorization(self.username, self.password),
+                ShipmentID=shipment_id,
+            )
+        except Fault as e:
```
```suggestion
        except zeep.exceptions.Fault as e:
```

sebalix commented on ebill_paynet/readme/INSTALL.rst:
```diff
@@ -0,0 +1,2 @@
+This module depends on `base_ebill_payment_contract` which is located on `account_invoicing`, so the OCA repository `account_invoicing` must also be present on the system.
```
```suggestion
This module depends on `base_ebill_payment_contract` which is located on `OCA/account-invoicing`, so this repository must also be present on the system.
```

TDu replied:
Changed

sebalix commented on ebill_paynet/views/ebill_payment_contract.xml:
```diff
@@ -0,0 +1,24 @@
+            ref="base_ebill_payment_contract.ebill_payment_contract_form_view"
+        />
+        <field name="arch" type="xml">
+            <field name="partner_id" position="after">
+                <field
+                    name="paynet_account_number"
+                    attrs="{'invisible': [('is_paynet_contract', '=', False)]}"
+                />
+                <field
+                    name="paynet_service_id"
+                    attrs="{'invisible': [('is_paynet_contract', '=', False)]}"
+                />
+                <field name="is_paynet_contract" invisible="1" />
+            </field>
```
I would put this in a separate group as `base_ebill_payment_contract` module could be used outside of Paynet context I suppose, so based on my comment on https://github.com/OCA/edi/pull/207#discussion_r475585357:
```suggestion
        <group name="base" position="after">
            <group name="paynet" position="after">
                <field
                    name="paynet_account_number"
                    attrs="{'invisible': [('is_paynet_contract', '=', False)]}"
                />
                <field
                    name="paynet_service_id"
                    attrs="{'invisible': [('is_paynet_contract', '=', False)]}"
                />
                <field name="is_paynet_contract" invisible="1" />
            </group>
        </group>
```

TDu replied:
Much better done

TDu committed:
[13.0][IMP] ebill_paynet

Fix the selection of the bank account of the sale order when paynet
transmit method is choosen.
Fix the get_active contract that has been moved to
base_ebill_payment_contract.
Add some fixes proposed by @sebalix.

TDu committed:
[13.0][IMP] ebill_paynet

Some improvements and fixes on invoice message generation :

* Fix the payment reference (it was using the old field pre v13)
* Allow to have a 2nd line in the biller name node, it uses the
company_name_suffix_1 on the company partner for it.
* Change the customer reference for the purchase linked to the invoice.
* Improve or so I hope the account number if it is a postal account the
  specific account number is used otherwise we use the bic.

yvaucher commented on ebill_paynet/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Ebill Paynet",
```
```suggestion
    "name": "eBill Paynet",
```

yvaucher commented on ebill_paynet/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,14 @@
+
+## Configure the contracts
+
+The contracts specific to ebilling are located in `Accounting - Customers - Ebill Payment Contract`
```
```suggestion
The contracts specific to e-billing are located in `Accounting - Customers - eBill Payment Contract`
```

yvaucher commented on ebill_paynet/views/message_template.xml:
```diff
@@ -0,0 +1,9 @@
+<odoo>
+    <template id="dws_reject_invoice">
+        <div>
+            <p>The Paynet system rejected the ebill that was send.</p>
```
```suggestion
            <p>The Paynet system rejected the eBill that was send.</p>
```

yvaucher commented on ebill_paynet/views/message_template.xml:
```diff
@@ -0,0 +1,9 @@
+    <template id="dws_reject_invoice">
+        <div>
+            <p>The Paynet system rejected the ebill that was send.</p>
+            <!-- <p><t t-esc="job_id" /></p> -->
```
A left over?

yvaucher reviewed (changes requested):
Small pre-review about terminology.

Please prefer `eBill` and `e-billing`

TDu committed:
fixup! [13.0][IMP] ebill_paynet

TDu committed:
[13.0][IMP] ebiil_paynet add qr

Implement the QR invoice functionality from SIX specifications.

Add on the ebill contract the choice of payment ESR or QR, so it can be
specific for each customer.
The XML payload has been updated. And the PDF data included improved, so
the correct PDF is generated depending of the type of payment.

TDu committed:
fixup! [13.0][IMP] ebiil_paynet add qr

TDu committed:
[13.0][IMP] ebiil_paynet B2B message

 Add the price and tax node for each line item.
 Improve the reference on the sale order.
 Add sale as a dependency and improve the tests by generating the
 invoice from a sale order.

TDu committed:
[13.0][IMP] ebiil_paynet

Fix naming of eBill
Improve import statement of zeep

TDu committed:
[13.0][IMP] ebiil_paynet B2B message fix tax on line

The specifcation accepts only one tax node by line.

TDu committed:
[13.0][FIX] ebiil_paynet - ESR bank account

The `l10n_ch_postal` field is used for bank account not belonging to the
company (like suppliers).
The `l10n_ch_isr_subscription_chf` field is the one used for bank account
belonging to the company.

TDu committed:
[13.0][IMP] ebill_paynet - Pdf sent

The actual invoice and the payment (ESR, QR) reports are not generated
together in one PDF so up to know only the page with the payment was
sent to Paynet.
Now both report are merge and sent to the system.

yvaucher commented on ebill_paynet/__manifest__.py:
```diff
@@ -0,0 +1,33 @@
+        "account_invoice_export",
+        "base_ebill_payment_contract",
+        "l10n_ch_base_bank",
+        "l10n_ch_qr",
```
This module has been renamed `l10n_ch_qriban`

TDu committed:
Fix url computed field on paynet service

Having this field stored will create problems for implementation
that uses different stacks (integration/production)
And it is not required

TDu committed:
Fix paynet message generation

Fix some line with a limited length of 35 char.
Round some decimal.

TDu committed:
Paynet fix the delivery address and party name

Use the partner_shipping_id information for the DELIVERY-PARTY node.
Also for all name nodes concerning a res.partner record, add a Line node
with the commercial company name if it is different to the partner name.

TDu committed:
Paynet fix fault handling on DWS communication

When downloading a shipment if a fault occurs on the communication with
the DWS. The module was trying to parse a content set to None.
So the exception must be raised and the information on the error will be
available in the job result.
Comment answered ;) and removed.

TDu committed:
fixup! Paynet fix fault handling on DWS communication

TDu committed:
Paynet fix invoicing line generated in XML

Remove invoice line with a display_type value. The filtering can be
overriden, if needed.
Handle properly the invoice line not linked to a product.
Fix some falsy values.

TDu committed:
Run pre-commit hooks

TDu committed:
Paynet switch Odoo order ref with customer order ref

Reading the specs it seemed to be the other way around.
But it has been confirmed by someone at SIX. That the reference for
linked sale order must be the customer order reference.

sebalix commented on ebill_paynet/models/paynet_invoice_message.py:
```diff
@@ -112,6 +112,7 @@ def _generate_payload(self):
         params = {
             "client_pid": self.service_id.client_pid,
             "invoice": self.invoice_id,
+            "invoice_line": self.invoice_id.paynet_invoice_line_ids(),
```
I suggest `invoice_lines`?

TDu replied:
Yes, fixed in next commit

TDu committed:
Paynet improve invoice currency and fix data length

Use the currency of the invoice to generated the message instead of
fixing it to CHF.
Fix the length of a name

TDu committed:
Fix the due date in the Paynet message

TDu committed:
Refactor ebill_paynet to allow for specific discount module

The payment term node in the Paynet message allows to add a discount
information in case of payment until a certain date.
As this feature does not exist in the standard implementation of Odoo,
this is extracted into another module, to give the option to use
different implementation for it.
See `ebill_paynet_account_financial_discount` for one of them.

TDu committed:
[13.0][ADD] ebill_paynet_account_financial_discount

This module implements the discount information in the payment terms node
of the Paynet message. In regards to the module `account_financial_discount`.

When the discount information is present in the payment terms of an invoice, they
will automatically be included in the payment term of the Paynet message.

TDu committed:
Fix linting

TDu committed:
Fix XML node content length with truncate

TDu committed:
Fix paynet communication with prod service

This fixes an issue when downloading a shipment from the DWS, while
using a production account.

TDu committed:
Fix module dependencies

The module l10n_ch_qr has been renamed to l10n_ch_qriban.

For this to work with an updated Odoo src the following PR is required:

* https://github.com/odoo/odoo/pull/57710

i-vyshnevska committed:
[FIX] remove customer fields (#3)

TDu commented:
I open a new PR https://github.com/OCA/l10n-switzerland/pull/585 see why in the description.
Thanks for all your reviews.

TDu commented:
Replaced by #585 and #586.
All comments have been taken care of.
